file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/LazyAttribute.kt

```diff
@@ -1,20 +1,63 @@
 package com.pkware.filesystem.attributecaching
 
-import kotlin.reflect.KProperty
-
 /**
- * Stores a mutable attribute that is lazily calculated via [initializer] whenever the value is `null` and accessed.
+ * Stores a mutable attribute [value] that is lazily calculated via [initializer] whenever the value is `null` and
+ * accessed.
+ *
+ * Can check the initialization status of the attribute via [initialized].
+ *
+ * If the mutable attribute [value] is set to `null` its [initialized] status is set to `false`.
+ *
  */
-internal class LazyAttribute<T>(private val initializer: () -> T?) {
+public class LazyAttribute<T>(private val initializer: () -> T?) {
 
-    private var value: T? = null
+    private var lazyValue: T? = null
 
-    operator fun getValue(thisRef: Any?, property: KProperty<*>): T? {
-        if (value == null) value = initializer()
-        return value
-    }
+    /**
+     * Shows whether this [LazyAttribute] is initialized or not. `true` for initialized, default is `false`.
+     */
+    public var initialized: Boolean = false
+        private set
+
+    /**
+     * The mutable attribute to store. If [initialized] is false the [initializer] function is called to set the
+     * attribute.
+     *
+     * The attribute can be `null`.
+     *
+     * If the mutable attribute is set to `null` its [initialized] status is set to `false`.
+     */
+    internal var value: T?
+        get() {
+            if (initialized) return lazyValue
+
+            lazyValue = initializer()
+            initialized = lazyValue != null
```

Comment:
```suggestion
            initialized = true
```

Is there ever a reason we would want to reload this?

Reply:
yes in the case where we set the attribute to null

Reply:
Two questions:

1. Why would ever want to set it to null?
2. If we set it to null, why wouldn't we want that value cached?

Reply:
if the initializer sets lazyValue to null somehow, that means nothing was cached and that initialized should remain false or be set to false

Reply:
thats up to the initializer itself, which could do anything, we have no control over what is defined in there

Reply:
We have **full** control over what's defined there. Nobody else is defining the initializer. We are. And the provider's `readAttributes` simply won't return null. And if it does so, it will do so deterministically. Thus, wasted retries.

Reply:
` val cachedDosAttributes = LazyAttribute {
        if (delegateSupportedFileAttributeViews.contains("dos")) {
            val dosFileAttributes = delegate.fileSystem.provider().readAttributes(
                delegate,
                DosFileAttributes::class.java,
            )
            dosFileAttributes
        } else {
            null
        }
    }`
    
  this is an initializer that returns null, and SHOULD return null if the view is not supported and thus we should record that nothing is cached/initialized.
  
  if somehow later the initializer is called again and the view is supported, we can cache it properly

Reply:
> if **somehow** later the initializer is called again and the view is supported, we can cache it properly

You should write a Star Wars movie. This can't happen.

Reply:
i can write a test with jimfs where it could happen, just swap filesystems with the same caching path

Reply:
copying a caching path across two machines with different filesystems/providers, some cached attributes could become null, some could become cached, and some just get copied

Reply:
> copying a caching path across two machines with different filesystems/providers, some cached attributes could become null, some could become cached, and **some just get copied while being cached** so they dont hit back at a filesystem operation.

So you're arguing against my proposal by using a scenario that would also be wonky under the current behavior?

Things like permissions/owner going across machines might be a huge problem and would likely error out.

Reply:
To be clear, I'm saying I think our happy path in the Unpacker would suffer any problems. We control the round trip. We know when things are loaded when they're assigned back.

But yeah, copy/move operations are a bit goofy. It's almost like they should reload the attributes for the sake of correctness, but I don't think our only customer would like us doing that.

Reply:
When would we be copying a cachingPath across two different file systems? As far as I know, we do our discovery & redaction "in-place", or use a virtual fileSystem (that should emulate the real one).

If this is a common or reasonably potential scenario, my thought is we'd actually want to throw an error due to the risk involved in that. Logic being: Since file attributes are considered important parts of the document, any risk of corrupting them violates our "do no harm" philosophy.

So either it's not a case we should worry about, or if we should worry about it, we might want to address it differently. But, I'm not (yet) seeing a reason to ever re-set the attributes.

Reply:
> When would we be copying a cachingPath across two different file systems? As far as I know, we do our discovery & redaction "in-place", or use a virtual fileSystem (that should emulate the real one).

In the case of PEM, the DiscoveryAgent is often on a different machine than the files it is operating on. Files are actually moving across two different underlying filesystems. This is a real concern.

However, we have the luxury of knowing the exact attributes we'd like to cache in the Unpacker. We load the attributes before modifying the file. Realistically, **we don't ever want** any results other than what was found when we initialized, hence the motivation for my feedback.

Are these assumptions currently a shortcoming of this library? Absolutely.
Did they already exist? Yes. 
Will my suggestion solve it? No.

But my suggestion will ensure correct behavior for the flow in which _we_ are currently using this library.

Reply:
> However, we have the luxury of knowing the exact attributes we'd like to cache in the Unpacker. We load the attributes before modifying the file, and those results are the ones we want cached. Realistically, we don't ever want any results other than what was found when we initialized, hence the motivation for my feedback, which would ensure that we never even bother to try loading more than once per path.

The single use case for this project right now is for our unpacker, however I wrote this file attribute cache for use beyond the unpacker when needed, because the scope of the initial design called for that.

We do in fact want results other than what was found when initialized when performing copy/move operations and the COPY_ATTRIBUTES flag is specified.

I added the extra copy attribute functionality to path conversion calls (absolutePath, normalize, and realPath) because I figured it's a nice bonus to copy attributes in those workflows and save some time when working with a file. I didnt initially realize the original implementation of this nice bonus was giving an other_iops hit when the cache was changed to be on-demand. With the current changes, I still kept some of the copy bonus in place and avoiding an other_iops hit to only copy currently cached/initialized attributes from the source path to the converted path.

The initialized variable, since it was introduced before the on demand cache, was never a one and done thing. It was always meant to be malleable and allow for attributes to be re-cached if they were initially null.

You can see that with even the pre-refactor changes for a given attribute as seen here:
`    /**
     * Shows whether [DosFileAttributes] are cached. `true` for cached, default is
     * `false`.
     */
    var dosAttributesCached = false
        private set

    private var dosAttributes by LazyAttribute {
        if (delegateSupportedFileAttributeViews.contains("dos")) {
            delegate.fileSystem.provider().readAttributes(delegate, DosFileAttributes::class.java)
            val dosFileAttributes = delegate.fileSystem.provider().readAttributes(
                delegate,
                DosFileAttributes::class.java,
            )
            dosAttributesCached = true
            dosFileAttributes
        } else {
            null
        }
    }`
  dosAttributesCached was never set to true in the case the view was not supported for this specific example

The entire motivation for this entire project was around to optimize other_iops performance with file attributes. It gets very tedious and frustrating to implement design changes here that have un-intended consequences on other_iops performance down the line.

Reply:
I should also add real support to setAttribute to bypass setting the delegate attribute to null and to set the cached attribute to null only (since calling the delegate setAttribute and setting that to null throws an NPE even though, the value can be nullable- doesn't make any sense from an implementation standpoint)